src/main/java/com/couchbase/lite/replicator/PusherInternal.java

```diff
@@ -295,12 +295,25 @@ public void beginReplicating() {
     @InterfaceAudience.Private
     public void changed(Database.ChangeEvent event) {
         final List<DocumentChange> changes = event.getChanges();
-        supportExecutor.submit(new Runnable() {
-            @Override
-            public void run() {
-                submitRevisions(changes);
-            }
-        });
+
+        // NOTE: REST API client can send POST/PUT /{db}/{doc} request too faster than
+        // push replicator can handle. If request comes from listener thread, submitting revision
+        // to batcher synchronusly. Otherwise, asynchronously.
+        // ref: https://github.com/couchbase/couchbase-lite-java-core/issues/1481
+        if (calledFromListener()) {
+            submitRevisions(changes);
+        } else {
+            supportExecutor.submit(new Runnable() {
+                @Override
+                public void run() {
+                    submitRevisions(changes);
+                }
+            });
+        }
+    }
+
```

Comment:
1. I added supportExecutor to avoid that caller is blocked by pause logic.
2. pause logic itself works fine. However, change events come faster, so changes and associated RevisionInternal is stuck in memory.
3. Native API could cause this issue too. 

Currently I don't have clean way to make REST client slow down the requests.

We could discuss this on Monday.